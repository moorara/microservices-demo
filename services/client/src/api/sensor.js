import axios from 'axios'

const http = axios.create({
  baseURL: '/api/v1/'
})

export default {
  async create (sensor) {
    try {
      const res = await http.post('sensors', sensor)
      return res.data
    } catch (err) {
      throw err
    }
  },

  async all (siteId) {
    try {
      const res = await http.get(`sensors?siteId=${siteId}`)
      return res.data
    } catch (err) {
      throw err
    }
  },

  async get (id) {
    try {
      const res = await http.get(`sensors/${id}`)
      return res.data
    } catch (err) {
      throw err
    }
  },

  async update (sensor) {
    try {
      const res = await http.put(`sensors/${sensor.id}`, sensor)
      return res.data
    } catch (err) {
      throw err
    }
  },

  async delete (id) {
    try {
      const res = await http.delete(`sensors/${id}`)
      return res.data
    } catch (err) {
      throw err
    }
  },
}
