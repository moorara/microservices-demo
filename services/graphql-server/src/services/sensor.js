const _ = require('lodash')

const Logger = require('../utils/logger')
const { createTracer } = require('../utils/tracer')

class SensorService {
  constructor (config, options) {
    options = options || {}
    this.logger = options.logger || new Logger('SensorService')
    this.histogram = options.histogram || { observe () {} }
    this.summary = options.summary || { observe () {} }
    this.tracer = options.tracer || createTracer({ serviceName: 'sensor-service' })

    this.store = {
      sensors: [
        { id: '1111-1111', siteId: 'aaaa-aaaa', name: 'temperature', unit: 'celsius', minSafe: -30.0, maxSafe: 30.0 },
        { id: '2222-2222', siteId: 'bbbb-bbbb', name: 'temperature', unit: 'fahrenheit', minSafe: -22.0, maxSafe: 86.0 }
      ]
    }
  }

  create (context, input) {
    const startTime = +new Date()
    const span = this.tracer.startSpan('create-sensor', { childOf: context.span })

    // TODO
    const sensor = Object.assign({}, input)
    sensor.id = _.uniqueId()
    this.store.sensors.push(sensor)

    const endTime = +new Date()
    const latency = (endTime - startTime) / 1000

    // Metrics
    const labelValues = { op: 'create_sensor', success: 'true' }
    this.histogram.observe(labelValues, latency)
    this.summary.observe(labelValues, latency)

    // Traces
    span.setTag('span.kind', 'client')
    span.setTag('peer.service', 'sensor-service')
    span.setTag('peer.address', 'sensor-service:4020')
    span.log({ event: 'create-sensor', message: '' })
    span.finish()

    return Promise.resolve(sensor)
  }

  all (context, siteId) {
    const startTime = +new Date()
    const span = this.tracer.startSpan('get-sensors', { childOf: context.span })

    // TODO
    const sensors = this.store.sensors.filter(s => s.siteId === siteId)

    const endTime = +new Date()
    const latency = (endTime - startTime) / 1000

    // Metrics
    const labelValues = { op: 'get_sensors', success: 'true' }
    this.histogram.observe(labelValues, latency)
    this.summary.observe(labelValues, latency)

    // Traces
    span.setTag('span.kind', 'client')
    span.setTag('peer.service', 'sensor-service')
    span.setTag('peer.address', 'sensor-service:4020')
    span.log({ event: 'get-sensors', message: '' })
    span.finish()

    return Promise.resolve(sensors)
  }

  get (context, id) {
    const startTime = +new Date()
    const span = this.tracer.startSpan('get-sensor', { childOf: context.span })

    // TODO
    const sensor = Object.assign({}, this.store.sensors.find(s => s.id === id))

    const endTime = +new Date()
    const latency = (endTime - startTime) / 1000

    // Metrics
    const labelValues = { op: 'get_sensor', success: 'true' }
    this.histogram.observe(labelValues, latency)
    this.summary.observe(labelValues, latency)

    // Traces
    span.setTag('span.kind', 'client')
    span.setTag('peer.service', 'sensor-service')
    span.setTag('peer.address', 'sensor-service:4020')
    span.log({ event: 'get-sensor', message: '' })
    span.finish()

    return Promise.resolve(sensor)
  }

  update (context, id, input) {
    const startTime = +new Date()
    const span = this.tracer.startSpan('update-sensor', { childOf: context.span })

    // TODO
    let updated
    const sensor = Object.assign({}, { id }, input)
    for (let i in this.store.sensors) {
      if (this.store.sensors[i].id === id) {
        this.store.sensors[i] = sensor
        updated = Object.assign({}, sensor)
        break
      }
    }

    const endTime = +new Date()
    const latency = (endTime - startTime) / 1000

    // Metrics
    const labelValues = { op: 'update_sensor', success: 'true' }
    this.histogram.observe(labelValues, latency)
    this.summary.observe(labelValues, latency)

    // Traces
    span.setTag('span.kind', 'client')
    span.setTag('peer.service', 'sensor-service')
    span.setTag('peer.address', 'sensor-service:4020')
    span.log({ event: 'update-sensor', message: '' })
    span.finish()

    return Promise.resolve(updated)
  }

  delete (context, id) {
    const startTime = +new Date()
    const span = this.tracer.startSpan('delete-sensor', { childOf: context.span })

    // TODO
    _.remove(this.store.sensors, s => s.id === id)

    const endTime = +new Date()
    const latency = (endTime - startTime) / 1000

    // Metrics
    const labelValues = { op: 'delete_sensor', success: 'true' }
    this.histogram.observe(labelValues, latency)
    this.summary.observe(labelValues, latency)

    // Traces
    span.setTag('span.kind', 'client')
    span.setTag('peer.service', 'sensor-service')
    span.setTag('peer.address', 'sensor-service:4020')
    span.log({ event: 'delete-sensor', message: '' })
    span.finish()

    return Promise.resolve()
  }
}

module.exports = SensorService
