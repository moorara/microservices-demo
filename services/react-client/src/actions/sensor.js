// https://redux.js.org/docs/advanced/AsyncActions.html

import sensorApi from '../api/sensor'

export const SITE_SENSORS_REQUESTED = 'SITE_SENSORS_REQUESTED'
export const SITE_SENSORS_RECEIVED = 'SITE_SENSORS_RECEIVED'
export const SITE_SENSORS_FAILED = 'SITE_SENSORS_FAILED'

export function requestSiteSensors (siteId) {
  return {
    type: SITE_SENSORS_REQUESTED,
    siteId
  }
}

export function receivedSiteSensors (sensors) {
  return {
    type: SITE_SENSORS_RECEIVED,
    sensors
  }
}

export function failedSiteSensors (err) {
  return {
    type: SITE_SENSORS_FAILED,
    err
  }
}

export function getSiteSensors (siteId) {
  return dispatch => {
    dispatch(requestSiteSensors(siteId))

    return sensorApi.all(siteId)
      .then(
        sensors => dispatch(receivedSiteSensors(sensors)),
        err => dispatch(failedSiteSensors(err))
      )
  }
}
