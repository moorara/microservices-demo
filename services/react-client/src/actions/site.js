// https://redux.js.org/docs/advanced/AsyncActions.html

import siteApi from '../api/site'

export const ALL_SITES_REQUESTED = 'ALL_SITES_REQUESTED'
export const ALL_SITES_RECEIVED = 'ALL_SITES_RECEIVED'
export const ALL_SITES_FAILED = 'ALL_SITES_FAILED'

export function requestAllSites () {
  return {
    type: ALL_SITES_REQUESTED
  }
}

export function receivedAllSites (sites) {
  return {
    type: ALL_SITES_RECEIVED,
    sites
  }
}

export function failedAllSites (err) {
  return {
    type: ALL_SITES_FAILED,
    err
  }
}

export function getAllSites () {
  return dispatch => {
    dispatch(requestAllSites())

    return siteApi.all()
      .then(
        sites => dispatch(receivedAllSites(sites)),
        err => dispatch(failedAllSites(err))
      )
  }
}
