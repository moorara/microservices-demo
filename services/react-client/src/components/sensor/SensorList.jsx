import React from 'react'
import PropTypes from 'prop-types'

import SensorListItem from './SensorListItem'

// Presentational Component
const SensorList = ({ sensors }) => (
  <table className="table is-bordered is-striped is-fullwidth is-hoverable">
    <thead>
      <tr>
        <th>Name</th>
        <th>Unit</th>
        <th>Minimum Safe Value</th>
        <th>Maximum Safe Value</th>
      </tr>
    </thead>
    <tbody>
      {sensors.map(sensor => <SensorListItem key={sensor.id} sensor={sensor} />)}
    </tbody>
  </table>
)

SensorList.propTypes = {
  sensors: PropTypes.array.isRequired
}

export default SensorList
