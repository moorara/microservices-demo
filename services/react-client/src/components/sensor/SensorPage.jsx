import React from 'react'

const SensorPage = (props) => (
  <section className="hero is-light is-large is-bold">
    <div className="hero-body">
      <div className="container">
        <h1 className="title">Sensors</h1>
        <div className="columns">
          <div className="column">
            Sensors ...
          </div>
        </div>
      </div>
    </div>
  </section>
)

export default SensorPage
