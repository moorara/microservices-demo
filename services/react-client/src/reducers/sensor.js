import * as actions from '../actions/sensor'

export const defaultState = {
  callsInProgress: 0,
  items: []
}

export default function (state = defaultState, action) {
  switch (action.type) {
    case actions.SITE_SENSORS_REQUESTED:
      return {
        ...state,
        callsInProgress: state.callsInProgress + 1
      }

    case actions.SITE_SENSORS_FAILED:
      return {
        ...state,
        callsInProgress: state.callsInProgress - 1
      }

    case actions.SITE_SENSORS_RECEIVED:
      return {
        items: action.sensors,
        callsInProgress: state.callsInProgress - 1
      }

    default:
      return state
  }
}
