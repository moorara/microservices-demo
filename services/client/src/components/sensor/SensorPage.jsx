import React from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

import SensorList from './SensorList'
import * as sensorActions from '../../actions/sensor'

// Container Component
export class SensorPage extends React.Component {
  static propTypes = {
    history: PropTypes.object.isRequired,
    match: PropTypes.object.isRequired,
    site: PropTypes.object.isRequired,
    sensors: PropTypes.array.isRequired,
    actions: PropTypes.object.isRequired,
  }

  constructor (props) {
    super(props)
    this.handleBack = this.handleBack.bind(this)
    this.handleAddSensor = this.handleAddSensor.bind(this)
  }

  handleBack () {
    this.props.history.goBack()
  }

  handleAddSensor () {
    console.log('Add Sensor clicked!')
  }

  // See https://reactjs.org/docs/react-component.html#shouldcomponentupdate
  shouldComponentUpdate (nextProps, nextState) {
    return true
  }

  componentDidMount () {
    const { actions } = this.props
    const siteId = this.props.match.params.id
    actions.getSiteSensors(siteId)
  }

  render () {
    const { site, sensors } = this.props
    return (
      <section className="hero is-light is-large is-bold">
        <div className="hero-body">
          <div className="container">
            <h1 className="title">{`${site.name} Sensors`}</h1>
            <SensorList sensors={sensors} />
            <input className="button is-link is-rounded" type="submit" value="Back" onClick={this.handleBack} />
            <span>&nbsp;&nbsp;</span>
            <input className="button is-danger is-rounded" type="submit" value="Add Sensor" onClick={this.handleAddSensor} />
          </div>
        </div>
      </section>
    )
  }
}

/*
 * Map store state to component props
 *   ownProps: the props passed to the connected component
 * See https://github.com/reactjs/react-redux/blob/master/docs/api.md
 */
export function mapStateToProps (state, ownProps) {
  const siteId = ownProps.match.params.id
  return {
    site: state.site.items.find(site => site.id === siteId) || {},
    sensors: state.sensor.items.filter(sensor => sensor.siteId === siteId)
  }
}

/*
 * Wrap each action creator with a dispatch call to store and pass actions to component props
 *   ownProps: the props passed to the connected component
 * See https://github.com/reactjs/react-redux/blob/master/docs/api.md
 * See https://redux.js.org/api-reference/bindactioncreators
 */
export function mapDispatchToProps (dispatch, ownProps) {
  return {
    actions: bindActionCreators(sensorActions, dispatch)
    /*
     * actions = {
     *   getSiteSensors: siteId => dispatch(sensorActions.getSiteSensors(siteId))
     * }
     */
  }
}

const connectToRedux = connect(mapStateToProps, mapDispatchToProps)

export default connectToRedux(withRouter(SensorPage))
