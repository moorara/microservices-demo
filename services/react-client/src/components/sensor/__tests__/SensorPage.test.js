import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import * as actions from '../../../actions/sensor'
import { SensorPage, mapStateToProps, mapDispatchToProps } from '../SensorPage'

describe('SensorPage', () => {
  describe('SensorPage', () => {
    let nextProps, nextState
    const props = {
      history: {},
      match: { params: {} },
      site: {},
      sensors: [],
      actions: {}
    }

    beforeEach(() => {
      props.history.goBack = jest.fn()
      props.actions.getSiteSensors = jest.fn()
    })

    test('renders connected react component with props', () => {
      props.match.params.id = '1111-1111'
      props.site = { id: '1111-1111', name: 'Station 1', location: 'Toronto, ON', priority: 2, tags: [ 'oxygen', 'hydrogen' ] }
      props.sensors = [
        { id: 'aaaa-aaaa', siteId: '1111-1111', name: 'temperature', unit: 'celsius', minSafe: -30, maxSafe: 30 }
      ]

      const wrapper = shallow(<SensorPage {...props} />)

      expect(shallowToJson(wrapper)).toMatchSnapshot()
      expect(wrapper.find('SensorList').prop('sensors')).toEqual(props.sensors)
      expect(wrapper.find('input').find({ value: 'Back' }).exists()).toBeTruthy()
      expect(wrapper.find('input').find({ value: 'Add Sensor' }).exists()).toBeTruthy()

      expect(wrapper.instance().shouldComponentUpdate(nextProps, nextState)).toBe(true)

      wrapper.find('input').find({ value: 'Back' }).simulate('click')
      expect(props.history.goBack).toHaveBeenCalled()
    })
    test('renders connected react component with props', () => {
      props.match.params.id = '2222-2222'
      props.sites = { id: '2222-2222', name: 'Station 2', location: 'Kingston, ON', priority: 3, tags: [ 'gold', 'silver' ] }
      props.sensors = [
        { id: 'bbbb-bbbb', siteId: '2222-2222', name: 'temperature', unit: 'fahrenheit', minSafe: -22, maxSafe: 86 }
      ]

      const wrapper = shallow(<SensorPage {...props} />)

      expect(shallowToJson(wrapper)).toMatchSnapshot()
      expect(wrapper.find('SensorList').prop('sensors')).toEqual(props.sensors)
      expect(wrapper.find('input').find({ value: 'Back' }).exists()).toBeTruthy()
      expect(wrapper.find('input').find({ value: 'Add Sensor' }).exists()).toBeTruthy()

      expect(wrapper.instance().shouldComponentUpdate(nextProps, nextState)).toBe(true)

      wrapper.find('input').find({ value: 'Back' }).simulate('click')
      expect(props.history.goBack).toHaveBeenCalled()
    })
  })

  describe('mapStateToProps', () => {
    const state = {
      site: {
        items: []
      },
      sensor: {
        items: []
      }
    }
    const ownProps = {
      match: {
        params: {}
      }
    }

    test('maps store state to component props', () => {
      ownProps.match.params.id = '1111-1111'
      state.site.items = [{ id: '1111-1111', name: 'Station 1', location: 'Toronto, ON', priority: 2, tags: [ 'oxygen', 'hydrogen' ] }]
      state.sensor.items = [
        { id: 'aaaa-aaaa', siteId: '1111-1111', name: 'temperature', unit: 'celsius', minSafe: -30, maxSafe: 30 }
      ]

      const props = mapStateToProps(state, ownProps)
      expect(props.site).toEqual(state.site.items[0])
      expect(props.sensors).toEqual(state.sensor.items)
    })
    test('maps store state to component props', () => {
      ownProps.match.params.id = '2222-2222'
      state.site.items = [{ id: '2222-2222', name: 'Station 2', location: 'Kingston, ON', priority: 3, tags: [ 'gold', 'silver' ] }]
      state.sensor.items = [
        { id: 'bbbb-bbbb', siteId: '2222-2222', name: 'temperature', unit: 'fahrenheit', minSafe: -22, maxSafe: 86 }
      ]

      const props = mapStateToProps(state, ownProps)
      expect(props.site).toEqual(state.site.items[0])
      expect(props.sensors).toEqual(state.sensor.items)
    })
  })

  describe('mapDispatchToProps', () => {
    let dispatch

    beforeEach(() => {
      dispatch = jest.fn()
      actions.getSiteSensors = jest.fn().mockReturnValue({
        type: 'GET_SITE_SENSORS'
      })
    })

    test('maps store dispatch and actions to component props', () => {
      const props = mapDispatchToProps(dispatch)
      props.actions.getSiteSensors('1111-1111')

      expect(actions.getSiteSensors).toHaveBeenCalledWith('1111-1111')
      expect(dispatch).toHaveBeenCalledWith({
        type: 'GET_SITE_SENSORS'
      })
    })
  })
})
