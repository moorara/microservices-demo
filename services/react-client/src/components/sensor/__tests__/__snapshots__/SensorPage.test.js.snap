// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SensorPage SensorPage renders connected react component with props 1`] = `
<section
  className="hero is-light is-large is-bold"
>
  <div
    className="hero-body"
  >
    <div
      className="container"
    >
      <h1
        className="title"
      >
        Station 1 Sensors
      </h1>
      <SensorList
        sensors={
          Array [
            Object {
              "id": "aaaa-aaaa",
              "maxSafe": 30,
              "minSafe": -30,
              "name": "temperature",
              "siteId": "1111-1111",
              "unit": "celsius",
            },
          ]
        }
      />
      <input
        className="button is-link is-rounded"
        onClick={[Function]}
        type="submit"
        value="Back"
      />
      <span>
          
      </span>
      <input
        className="button is-danger is-rounded"
        onClick={[Function]}
        type="submit"
        value="Add Sensor"
      />
    </div>
  </div>
</section>
`;

exports[`SensorPage SensorPage renders connected react component with props 2`] = `
<section
  className="hero is-light is-large is-bold"
>
  <div
    className="hero-body"
  >
    <div
      className="container"
    >
      <h1
        className="title"
      >
        Station 1 Sensors
      </h1>
      <SensorList
        sensors={
          Array [
            Object {
              "id": "bbbb-bbbb",
              "maxSafe": 86,
              "minSafe": -22,
              "name": "temperature",
              "siteId": "2222-2222",
              "unit": "fahrenheit",
            },
          ]
        }
      />
      <input
        className="button is-link is-rounded"
        onClick={[Function]}
        type="submit"
        value="Back"
      />
      <span>
          
      </span>
      <input
        className="button is-danger is-rounded"
        onClick={[Function]}
        type="submit"
        value="Add Sensor"
      />
    </div>
  </div>
</section>
`;
